neo/Consensus/ConsensusService.cs

```diff
@@ -289,18 +361,30 @@ private void OnPrepareRequestReceived(ConsensusPayload payload, PrepareRequest m
 
         private void OnPrepareResponseReceived(ConsensusPayload payload, PrepareResponse message)
         {
-            if (context.Signatures[payload.ValidatorIndex] != null) return;
+            if (context.State.HasFlag(ConsensusState.CommitSent) && context.State.HasFlag(ConsensusState.SignatureSent)) return;
             Log($"{nameof(OnPrepareResponseReceived)}: height={payload.BlockIndex} view={message.ViewNumber} index={payload.ValidatorIndex}");
-            byte[] hashData = context.MakeHeader()?.GetHashData();
-            if (hashData == null)
-            {
-                context.Signatures[payload.ValidatorIndex] = message.Signature;
-            }
-            else if (Crypto.Default.VerifySignature(hashData, message.Signature, context.Validators[payload.ValidatorIndex].EncodePoint(false)))
+
+
+            //====================================================================
+            //The following is like an additional feature 
+            // In the original code we were saving the signature and later verifying.
+            // Here we can check the PreparePayload if it did not arrive, because all PrepareREsponse Payloads carries that
+            if (context.PreparePayload == null)
             {
-                context.Signatures[payload.ValidatorIndex] = message.Signature;
-                CheckSignatures();
+                // We need to check if the Node send the orrect PreparePayload from the expected PrimaryIndex
```

Comment:
I don't like very much send twice the BlockHeaders ... but i understand it

But.. why we need `PrepareRequest`? could be a PrepareResponse but only from the leader, now i can't see the difference

Reply:
The BlockHeaders are not sent twice, only one `PrepareRequest` per payload.
The additional feature is this aforementioned payload merged with the `PrepareResponse payload`, which has the potential speeding up the process for when PrepareRequest did not arrive.

In fact, the important feature was to have a local variable with the `PrepareRequest payload`. This variable on `context` is the one that is used for regeneration of other stuck nodes.

It was suggested by @erikzhang in our previous conversation over the commit phase.
At the end, I understand you, @shargon, but I think it is a speed up on the broadcast that has several long-term benefits when the system is even more decentralized.